src/main/java/mate/academy/spring/service/impl/CustomUserDetailsServiceImpl.java

```diff
@@ -0,0 +1,39 @@
+package mate.academy.spring.service.impl;
+
+import static org.springframework.security.core.userdetails.User.withUsername;
+
+import java.util.Optional;
+import mate.academy.spring.model.Role;
+import mate.academy.spring.model.User;
+import mate.academy.spring.service.UserService;
+import org.springframework.security.core.userdetails.User.UserBuilder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CustomUserDetailsServiceImpl implements UserDetailsService {
+    private final UserService userService;
+
+    public CustomUserDetailsServiceImpl(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username)
+            throws UsernameNotFoundException {
+        Optional<User> optionalUser = userService.findByEmail(username);
+        if (optionalUser.isEmpty()) {
+            throw new UsernameNotFoundException("User not found");
+        }
+        User user = optionalUser.orElseThrow();
+        UserBuilder userBuilder =
+                withUsername(username);
+        userBuilder.password(user.getPassword());
+        userBuilder.authorities(user.getRoles().stream()
+                .map(Role::getRoleName)
+                .toArray(String[]::new));
+        return userBuilder.build();
+    }
```

Comment:
Usually we are using builder in such way:
```suggestion
        String[] roles = user.getRoles().stream()
                .map(Role::getRoleName)
                .toArray(String[]::new));
        return userBuilder
                .username(username)
                .password(user.getPassword())
                .authorities(roles)
                .build();
    }
```